backend/packages/Upgrade/src/api/services/ExperimentUserService.ts

```diff
@@ -33,32 +35,59 @@ export class ExperimentUserService {
     return this.userRepository.findOne({ id });
   }
 
-  public async create(users: Array<Partial<ExperimentUser>>, logger: UpgradeLogger): Promise<ExperimentUser[]> {
-    logger.info({ message: 'Create a new User. Metadata of the user =>', details: users });
-    // TODO: Pratik please review this eslint error, is this working as intended?
-    const multipleUsers = users.map((user) => {
-      // eslint-disable-next-line no-self-assign
-      user.id = user.id;
-      return user;
+  public async upsertOnChange(
+    oldExperimentUser: RequestedExperimentUser,
+    newExperimentUser: Partial<ExperimentUser>,
+    logger: UpgradeLogger
+  ): Promise<ExperimentUser[]> {
+    if (!oldExperimentUser) {
+      return this.create([newExperimentUser], logger);
+    }
+
+    // Check if either the group or workingGroup is different or not present
+    const isGroupAbsent = !oldExperimentUser?.group && !newExperimentUser?.group;
+    const isWorkingGroupAbsent = !oldExperimentUser?.workingGroup && !newExperimentUser?.workingGroup;
+
+    // Do not call create if both group and workingGroup are absent
+    if (isGroupAbsent && isWorkingGroupAbsent) {
+      return [oldExperimentUser];
+    }
+
+    // sort the group membership values
+    const oldGroupKeys = Object.keys(oldExperimentUser.group);
+    const newGroupKeys = Object.keys(newExperimentUser.group);
+
+    oldGroupKeys.forEach((key) => {
+      oldExperimentUser.group[key].sort();
     });
-    // insert or update in the database
-    const updatedUsers = await this.userRepository.save(multipleUsers);
+    newGroupKeys.forEach((key) => {
+      newExperimentUser.group[key].sort();
+    });
+
+    const isWorkingGroupSame = isEqual(oldExperimentUser.workingGroup, newExperimentUser.workingGroup);
+    const isSame = isEqual(oldExperimentUser.group, newExperimentUser.group) && isWorkingGroupSame;
```

Comment:
Can we name the `isSame` to something more specific. I guess `isGroupAndWorkingGroupSame` might be good one.

Reply:
code updated